### PR TITLE
Add synthetic AssignmentNodes for ternary expressions

### PR DESCRIPTION
**Pull request by d367wang**
typetools PR 5000 https://github.com/eisop/checker-framework/pull/151/commits/48f2652bc8bf4801b2e750cd92325583939f2f52 added synthetic variables for ternary expressions to the CFG.
This broke how the Nullness Checker handles ternary expressions, leading to false positives.

This PR fixes handling of ternary expressions by using a special AssignmentNode that does not merge stores, for the assignment of the synthetic variable used in a ternary expression.

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 1]
```diff
@@ -2610,7 +2610,6 @@ private void extendWithAssignmentForConditionalExpr(
         handleArtificialTree(condExprVarUseTree);
         LocalVariableNode condExprVarUseNode = new LocalVariableNode(condExprVarUseTree);
         condExprVarUseNode.setInSource(false);
-        extendWithNode(condExprVarUseNode);
```
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2613]: Should also change the method name and doc?

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2613]: Yes, please. The test is failing, so maybe this is needed somewhere?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AssignmentNode.java [hunk 2]
```diff
@@ -62,14 +88,19 @@ public Tree getTree() {
         return tree;
     }
 
+    /** Check if the then-store and else-store should be merged. */
```
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AssignmentNode.java:91]:
```suggestion
    /** Check if the assignment node is synthetic, e.g. synthetic assignment in a ternary expression. */
```

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 3]
```diff
@@ -842,9 +853,18 @@ protected void addInformationFromPreconditions(
                }
         */
 
-        processCommonAssignment(in, lhs, rhs, store, rhsValue);
-
-        return new RegularTransferResult<>(finishValue(rhsValue, store), store);
+        if (n.isSynthetic() && in.containsTwoStores()) {
```
Comment (d367wang) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:856]:
```suggestion
        if (n.isSynthetic() && in.containsTwoStores()) {
        // This is a synthetic assignment node created for ternary expression. In this case the `then` and `else` store are not merged.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 4]
```diff
@@ -2586,31 +2582,33 @@ public Node visitConditionalExpression(ConditionalExpressionTree tree, Void p) {
...
         AssignmentNode assignmentNode =
-                new AssignmentNode(assign, treeAndLocalVarNode.second, caseExprNode);
+                new AssignmentNode(assign, treeAndLocalVarNode.second, caseExprNode, true);
         assignmentNode.setInSource(false);
         extendWithNode(assignmentNode);
     }
 
     /**
-     * Extend the CFG with a {@link LocalVariableNode} representing a use of some variable
+     * Build a pair of <{@link IdentifierTree} and {@link LocalVariableNode} to represent a use of
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2599]:
```suggestion
     * Build a pair of {@link IdentifierTree} and {@link LocalVariableNode} to represent a use of
```

**Pull request merged**